scripts/merge_deps/main.go

```diff
@@ -0,0 +1,214 @@
+	moduleFileName      = "go.mod"
+	backupDirName       = "tmp"
+)
+
+func main() {
+	if len(os.Args) != 4 {
+		fmt.Printf("Must provide 3 arguments: \n\t- plugin go.mod file path \n\t- Glooe go.mod file path\n\t- merge attempts plugin go.mod file\n")
+		os.Exit(1)
+	}
+
+	pluginsDependenciesFilePath := os.Args[1]
+	glooDependenciesFilePath := os.Args[2]
+	var (
+		mergeAttempt    int
+		nonMatchingDeps []checks.DependencyInfoPair
+		moduleInfo      *checks.ModuleInfo
+		err             error
+	)
+	if mergeAttempt, err = strconv.Atoi(os.Args[3]); err != nil {
+		fmt.Printf("Provided 3th arguments is not a number\n")
+		os.Exit(1)
+	}
+
+	if moduleInfo, err = checks.ParseModuleInfo(moduleFileName); err != nil {
+		fmt.Printf("Failed to read plugin module info: %s/n", err.Error())
+		os.Exit(1)
+	}
+
+	var mergedDeps map[string]checks.DependencyInfo
+
+	for i := 1; mergeAttempt > 0 && i <= mergeAttempt; i++ {
+		if nonMatchingDeps, err = checks.CompareDependencies(pluginsDependenciesFilePath, glooDependenciesFilePath); err != nil {
+			fmt.Printf("Failed to compare dependencies: %s/n", err.Error())
+			os.Exit(1)
+		}
+
+		if len(nonMatchingDeps) == 0 {
+			fmt.Println("All shared dependencies match")
+			os.Exit(0)
+		}
+		fmt.Println("Plugin and Gloo Enterprise dependencies do not match!")
+		if i < mergeAttempt {
+			fmt.Printf("Adjusting Plugin dependencies and start comparing again (%d times)\n", i)
+		}
+
+		if mergedDeps, err = mergeDependencies(pluginsDependenciesFilePath, nonMatchingDeps); err != nil {
+			fmt.Printf("Failed to merge non matching dependencies: %s/n", err.Error())
+			os.Exit(1)
+		}
+
+		if err = backupPluginModuleFile(i); err != nil {
+			fmt.Printf("Failed to backup current '%s' file: %s/n", moduleFileName, err.Error())
+			os.Exit(1)
+		}
+
+		if err = createPluginModuleFile(moduleInfo, mergedDeps); err != nil {
+			fmt.Printf("Failed to write new merged '%s' file: %s/n", moduleFileName, err.Error())
+			os.Exit(1)
+		}
+	}
+
+	// 1. Write the report to stdout
+	reportBytes, err := json.MarshalIndent(nonMatchingDeps, "", "  ")
+	if err != nil {
+		fmt.Printf("Failed to marshall error report to JSON: %s/n", err.Error())
+		os.Exit(1)
+	}
+	fmt.Println(string(reportBytes))
+
+	// 2. Write the report to a file
+	fmt.Printf("Writing error report file [%s]\n", errorReportFile)
+	if err := ioutil.WriteFile(errorReportFile, reportBytes, 0644); err != nil {
+		fmt.Printf("Failed to write error report file: %s/n", err.Error())
+	}
+
+	// 3. Create a file with suggested changes to go.mod
+	if err := createSuggestionsFile(nonMatchingDeps); err != nil {
+		fmt.Printf("Failed to create suggestions file: %s/n", err.Error())
+	}
+	os.Exit(1)
+}
+
+func backupPluginModuleFile(suffix int) error {
+	if err := os.MkdirAll(backupDirName, os.ModePerm); err != nil {
+		return err
+	}
+	backupModuleFileName := fmt.Sprintf("%s/%s-%d", backupDirName, moduleFileName, suffix)
+	return os.Rename(moduleFileName, backupModuleFileName)
```

Comment:
This approach makes it pretty awkward to run / iterate on the script from a debugger in an IDE, because it will try to reload the go.mod file after it has been moved. Also, it means if this merge process doesn't succeed, then you may be left with a broken working directory (i.e. you may need to revert the go.mod changes to get your IDE happy again, or to make the next script invocation work). 

For these reasons, I think it would be preferable to have the script output the entire contents of a new file (maybe to your tmp directory) and then after the script completes successfully, and any additional verification is performed on the go.mod file, swap out the contents.

Reply:
This script can also know to stop when a merge attempts results in no changes to the output go.mod.

Reply:
For context, I am testing by running:
`GLOOE_VERSION=1.4.0-beta5 make get-glooe-info -B` 

Then running `make merge-deps`, or trying to debug that in the IDE. I'm not seeing it get past this suggestion:

```
[
  {
    "message": "Please pin your dependency to the same version as the Gloo one using a [require] clause",
    "pluginDependencies": {
      "name": "golang.org/x/tools",
      "version": "v0.0.0-20200423204450-38a97e00a8a1",
      "replacement": false
    },
    "glooDependencies": {
      "name": "golang.org/x/tools",
      "version": "v0.0.0-20200423205358-59e73619c742",
      "replacement": false
    }
  }
]
```

Reply:
I've refactored the PR which now passes the suggestions.

Regarding your debugging comment:
I've changed the code to (as you suggested) save the suggested module in the temp dir and once all deps match, copy it to the `go.mod`